docs/design/rosbag2_playback_time.md

```diff
@@ -0,0 +1,200 @@
+For playback we want to support the following "time-control" features:
+* Pause and resume time
+  * When time is resumed, rosbag2 will play the next message in sequence after the last one it published
+  * There is an alternate concept of "pause and resume" that means "suppress recording/publishing while time continues, such that messages are skipped" - I use "pause and resume" in this document in the sense of "stopping and starting the flow of time". An alternative design handles that other feature, which may be more appropriately named something else such as "suppress", or "mute" for an analogy with an audio playback.
+* Set a (forward) rate of playback - either faster or slower than real time
+  * This is already implemented as of this writing, but this design proposes a change in the implementation
+* Jump back or forward to an arbitrary point in time
+  * This is called "jump" in the time interface, but will likely be exposed in the player interface as "seek"
+
+These are the time situations to handle:
+1. Steady Time - Rosbag2 keeps time internally with reference to a monotonic system clock, and neither publishes or subscribes to `/clock`
+  * This is the default behavior
+2. Rosbag2 acts as ROS Time Source by publishing to the `/clock` topic (a simple extension of case 1)
+  * Chosen by `--clock` option to `ros2 bag play`
+3. Rosbag2 is driven by an external ROS Time Source - most commonly the Gazebo simulator
+  * In this case, Rosbag2 cannot use "time-control" features presented above, since it is a passive consumer of time
+  * Chosen by `--use-sim-time` argument to `ros2 bag play`
+
+NOTE: The user is responsible for ensuring that there is only one publisher on `/clock` - Rosbag2 will not try to de-conflict multiple ROS Time Sources, but it will print a warning if more than one publisher is detected.
+
+## Notes on current implementation (as of writing this design)
+
+`rosbag2_transport::Player` currently uses `std::chrono::system_clock` to query time, and `std::this_thread::sleep_until` to wait between publishing messages, with explicit handling of playback rate.
+
+## Proposal
+
+Pass a `PlayerClock` instance to `rosbag2_transport::Player`
+* Use `PlayerClock::now` to query starting time
+* Use `PlayerClock::sleep_until` between messages
+  * `rclcpp::Clock` does not yet implement `sleep_until` - as noted in https://github.com/ros2/rcl/issues/898
+  * If we are not able to contribute this feature into `rclcpp` in time for the Galactic API freeze - it may need to temporarily go into a rosbag2-based subclass for the Galactic release.
+
+```
+/**
+ * Used to control the timing of bag playback.
+ * This clock should be used to query times and sleep between message playing,
+ * so that the complexity involved around time control and time sources
+ * is encapsulated in this one place.
+ * Internally, it may own an rclcpp::Clock, but does not override the class in order to implement
+ * a slightly different API.
+ */
+class PlayerClock
```

Comment:
I'd generally be hesitant to put in larger amount of source code in a design doc. It will eventually be out of sync with the actual implementation.

Reply:
Could we rather create some sort of UML or simple diagram to visualize it?

Reply:
I was trying to leave all implementation out while suggesting an API - but you're right, it is likely to get out of date. UML still suggests an API... but would it seem less final that way?

Reply:
it's not a deal breaker, but I'd rather highlight the key concepts here (maybe even just pseudo code) and leave out stuff like `virtual ~PlayerClock()`

Reply:
That's fair - I will re-work it a little bit to be less prescriptive